website/www/site/content/en/blog/deploy-python-pipeline-on-flink-runner.md

```diff
@@ -0,0 +1,852 @@
+
+{{< highlight py >}}
+# beam/word_len/run.py
+from . import *
+
+run()
+{{< /highlight >}}
+
+Overall, the pipeline package is structured as shown below.
+
+{{< highlight bash >}}
+tree beam/word_len
+
+beam/word_len
+├── __init__.py
+├── run.py
+└── word_len.py
+{{< /highlight >}}
+
+### Build Docker Images
+
+As mentioned earlier, we build a custom Docker image (*beam-python-example:1.16*) that is used for deploying a Flink cluster as well as for executing Java user code of the *Kafka Connector I/O*.
+
+```Dockerfile
+# beam/Dockerfile
+FROM flink:1.16
+
+COPY --from=apache/beam_java11_sdk:2.56.0 /opt/apache/beam/ /opt/apache/beam/
+```
+
+We also build another custom Docker image (*beam-python-harness:2.56.0*) to run Python user code (*SDK Harness*). From the Python SDK Docker image, it first installs JDK Development Kit (JDK) and downloads the *Java IO Expansion Service* Jar file. Then, Beam pipeline packages are copied to the */app* folder, and it ends up adding the app folder into the *PYTHONPATH* environment variable, which makes the packages to be searchable.
+
+```Dockerfile
+# beam/Dockerfile-python-harness
+FROM apache/beam_python3.10_sdk:2.56.0
+
+ARG BEAM_VERSION
+ENV BEAM_VERSION=${BEAM_VERSION:-2.56.0}
+ENV REPO_BASE_URL=https://repo1.maven.org/maven2/org/apache/beam
+
+RUN apt-get update && apt-get install -y default-jdk
+
+RUN mkdir -p /opt/apache/beam/jars \
+  && wget ${REPO_BASE_URL}/beam-sdks-java-io-expansion-service/${BEAM_VERSION}/beam-sdks-java-io-expansion-service-${BEAM_VERSION}.jar \
+          --progress=bar:force:noscroll -O /opt/apache/beam/jars/beam-sdks-java-io-expansion-service.jar
+
+COPY word_len /app/word_len
+COPY word_count /app/word_count
+
+ENV PYTHONPATH="$PYTHONPATH:/app"
+```
+
+As the custom images should be accessible in the minikube cluster, we should point the terminal's docker-cli to the minikube's Docker engine. Then, the images can be built as usual using `docker build`.
+
+{{< highlight bash >}}
+eval $(minikube docker-env)
+docker build -t beam-python-example:1.16 beam/
+docker build -t beam-python-harness:2.56.0 -f beam/Dockerfile-python-harness beam/
+{{< /highlight >}}
+
+## Deploy Stream Processing App
+
+The Beam pipeline is executed on a [Flink session cluster](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/custom-resource/overview/#session-cluster-deployments), which is deployed via the Flink Kubernetes Operator. Note that the [application deployment mode](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/custom-resource/overview/#application-deployments) where the Beam pipeline is deployed as a Flink job doesn't seem to work (or I don't understand how to do so yet) due to either job submission timeout error or failing to upload the job artifact. After the pipeline is deployed, we check the output of the application by sending text messages to the input Kafka topic.
```

Comment:
```suggestion
The Beam pipeline is executed on a [Flink session cluster](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/custom-resource/overview/#session-cluster-deployments), which is deployed by the Flink Kubernetes Operator. The [application deployment mode](https://nightlies.apache.org/flink/flink-kubernetes-operator-docs-main/docs/custom-resource/overview/#application-deployments) where the Beam pipeline is deployed as a Flink job doesn't seem to work (or I don't understand how to do so yet) due to either a job submission timeout error or a failure to upload the job artifact. After the pipeline is deployed, we check the output of the application by sending text messages to the input Kafka topic.
```

Reply:
For the notes about things that aren't working, it might be better to put them in a Note callout rather than embedded in the text.